### DISCUSSION TITLE
Problems with Basic Transformer and linear solver LAPACK

### DISCUSSION THREAD

**Discussion by kmerbcn** [OpenModelica/OpenModelica#12493] [General]
Hello

I am stucked with the Basic Transformer model. It seems that as soon as I try to solve a model with a basic transformer and a load resistor, the model becomes unsolvable.
It seems that the linear solver seems the matrix with lower rank than the number of equations.
It is very strange because the model is an entirely copy of the example "CompareTransformers" and that model works in the example.
What am I doing wrong?

The log file states as 

C:/Users/jgarcia/AppData/Local/Temp/OpenModelica/OMEdit/PruebaTrafo/PruebaTrafo.exe -port=63206 -logFormat=xmltcp -override=startTime=0,stopTime=50,stepSize=0.001,tolerance=1e-3,solver=dassl,outputFormat=mat,variableFilter=.* -r=C:/Users/jgarcia/AppData/Local/Temp/OpenModelica/OMEdit/PruebaTrafo/PruebaTrafo_res.mat -maxIntegrationOrder=1 -initialStepSize=1e-9 -maxStepSize=1e-6 -w -lv=LOG_STDOUT,LOG_ASSERT,LOG_LS,LOG_STATS -inputPath=C:/Users/jgarcia/AppData/Local/Temp/OpenModelica/OMEdit/PruebaTrafo -outputPath=C:/Users/jgarcia/AppData/Local/Temp/OpenModelica/OMEdit/PruebaTrafo
initialize linear system solvers
2 linear systems
Start solving Linear System 19 (size 1) at time 0 with Lapack Solver
Failed to solve linear system of equations (no. 19) at time 0.000000, system is singular for U[2, 2].
Matrix U
        -0 
Output vector x
[ 1]                 -0.1
The default linear solver fails, the fallback solver with total pivoting is started at time 0.000000. That might raise performance issues, for more information use -lv LOG_LS.
Start solving Linear System 19 (size 1) at time 0 with Total Pivot Solver
Total Pivot: Matrix (nearly) singular at initialization.
Continuing anyway. For more information please use -lv LOG_LS.
rank =  0
under-determined linear system not solvable!
Error solving linear system of equations (no. 19) at time 0.000000.
Solving linear system 19 fails at time 0. For more information use -lv LOG_LS.
[1] Real der(basicTransformer.i2)(start=0, nominal=1000)
Solving linear system 19 failed at time=0. For more information please use -lv LOG_LS.
simulation terminated by an assertion at initialization

The model looks as follows 

![2024-06-03 11_33_33-OMEdit - OpenModelica Connection Editor](https://github.com/OpenModelica/OpenModelica/assets/62482095/8b4cdba5-0013-4f24-ae7f-50bdd55b905f)

The problem comes with the Load Resistor, Rl, as soon as I remove the load, the model works. It seems the model has a problem with the secondary current, which i can't see due to the simplicity of the model

**Comment by casella** [comment]
Please post the .mo file of your example. Maybe there is some problem in how you set the component parameters, are you sure they are all correct?

**Reply by kmerbcn** [reply]
It seems that I can't upload model files. How can I post it?
Thanks

**Comment by kmerbcn** [comment]
[PruebaTrafo.zip](https://github.com/user-attachments/files/15532115/PruebaTrafo.zip)

**Comment by casella** [comment]
@kmerbcn, if you remove the load, the secondary circuit is opened, so the system becomes trivial. However, if you connect resistor12 to ground, the problem is still there.

As I understand, the chosen set of parameters for the transformer (L1 = 0.16, L2 = 0.04, M = 0.08) makes the impedance matrix
```
| L1   M|
| M   L2|
```
singular (the determinant is zero), so the model cannot be solved. In the original model, the parameters are slightly different from those values, so the system is well-posed.

**Reply by kmerbcn** [reply]
Dear Francesco

You are right the rank of the matrix is 1. Nevertheless, this example comes directily from a copy and paste from the Example in the electrical library CompareTransformes.mo . How could you overcome the situation? 
Would you slightly change the parameters?

**Comment by casella** [comment]
The example comes from there, but the parameters of the CompareTransformers model are defined as:
```modelica
  parameter SI.Resistance R1=0.01  "Primary resistance w.r.t. primary side";
  parameter SI.Inductance L1sigma=0.05/(2*pi*f)  "Primary leakage inductance w.r.t. primary side";
  parameter SI.Inductance Lm1= 10./(2*pi*f)  "Magnetizing inductance w.r.t. primary side";
  parameter SI.Inductance L2sigma=0.05/(2*pi*f)/n^2  "Secondary leakage inductance w.r.t. secondary side";
  parameter SI.Resistance R2=0.01/n^2  "Secondary resistance w.r.t. secondary side";
  parameter SI.Resistance RL=1/n^2 "Load resistance";
  final parameter SI.Inductance L1=L1sigma + M*n  "Primary no-load inductance";
  final parameter SI.Inductance L2=L2sigma + M/n  "Secondary no-load inductance";
  final parameter SI.Inductance M=Lm1/n "Mutual inductance";
```
which are not the same as in your example. Numerically, L1 = 0.15995, L2 = 0.03998, M = 0.0795, which lead to a non-zero jacobian. Why those values where selected for the example, I have no idea; in any case, the goal of the example is to compare two different (but equivalent) models of the transformer.

**Reply by kmerbcn** [reply]
Hello Francesco

I understand why the user put these numbers. Bascially, he "added" some noise to avoid numerical issues while keeping results still acceptable with the reality. I was playing with that approach before.

Another question I have is whether you can avoid scaling in Lapack solver. I think then transient and non-linear solver would work better. Do you reccomend to use an implicit solver such as Backward Euler?

Br,

Javier Martinez

**Reply by casella** [reply]
> Another question I have is whether you can avoid scaling in Lapack solver. I think then transient and non-linear solver would work better.

In my experience scaling with nominal values is always beneficial, see http://dx.doi.org/10.1145/3158191.3158192

> Do you recommend to use an implicit solver such as Backward Euler?

DASSL (variable step-size BDF) usually works well, except for systems with very fast and very poorly damped eigenvalues, in which case RADAU is more efficient.

In any case, the problem here is not with the ODE solver, but rather with the transformation from DAE to explicit ODE, which uses an algebraic solver (Lapack).
> 
> Br,
> 
> Javier Martinez

**Reply by kmerbcn** [reply]
Thanks a lot for the reference. I will have a look.
Concerning to your last point  (transformation from DAE to explicit ODE), I was reading in other posts concerning that topic.
I read that you need to add a compiler flag --daemode . Using scripting I can do it, but I am now using the connector editor.
Where do you add the compiler flag? 
I tried as follows  :   
Simulation --> Simulation Setup --> C/C++ Compiler Flags --> --daemode 
I push OK and the error message is  :  

C:/Program Files/OpenModelica1.22.3-64bit/share/omc/scripts/Compile.bat PruebaTransformador gcc mingw64 parallel dynamic 1 0
PATH = "C:\PROGRA~1\OPENMO~1.3-6\tools\msys\mingw64\bin;C:\PROGRA~1\OPENMO~1.3-6\tools\msys\mingw64\bin\..\..\usr\bin;"
mingw32-make: Entering directory 'C:/Users/Javito/AppData/Local/Temp/OPENMO~1/OMEdit/PRBBEF~1'
clang -municode  -Os -DOM_HAVE_PTHREADS -Wno-parentheses-equality -falign-functions -mstackrealign -msse2 -mfpmath=sse   --daemode  -I"C:/Program Files/OpenModelica1.22.3-64bit/include/omc/c" -I"C:/Program Files/OpenModelica1.22.3-64bit/include/omc" -I. -DOPENMODELICA_XML_FROM_FILE_AT_RUNTIME -DOMC_MODEL_PREFIX=PruebaTransformador -DOMC_NUM_MIXED_SYSTEMS=0 -DOMC_NUM_LINEAR_SYSTEMS=2 -DOMC_NUM_NONLINEAR_SYSTEMS=0 -DOMC_NDELAY_EXPRESSIONS=0 -DOMC_NVAR_STRING=0  -c -o PruebaTransformador.o PruebaTransformador.c
clang: error: unsupported option '--daemode'
mingw32-make: *** [<builtin>: PruebaTransformador.o] Error 1
mingw32-make: Leaving directory 'C:/Users/Javito/AppData/Local/Temp/OPENMO~1/OMEdit/PRBBEF~1'
Compilation process failed. Exited with code 0x00000002.

What is wrong?

**Comment by casella** [comment]
`--daeMode` is meant to be used if you _don't_ want to turn DAEs into ODEs and then integrate them as ODEs, but rather integrate them directly as DAEs. Currently only available if you put `--daeMode` in the Additional Translation Flags field of Translation Flags

**Reply by kmerbcn** [reply]
Hello Francesco,

Thanks for the reply. How do you enable this compiler flag in the GUI interface? In windows I have problems when I put it into the simulation parameters. As stated above, the compiler states :

clang error 

It seems the compiler flag, --daemode, does not like it

I also changed the solver to IDA

**Reply by casella** [reply]
![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/14fab76c-ece3-4bcd-ad47-fa7fc34a013e)

The solver is switched to IDA automatically, since it is the only DAE solver we implemented.

**Comment by kmerbcn** [comment]
Dear Francesco. Thanks for the clarification. I was writing the flag in the first tab -->General and compiler flags

**Reply by kmerbcn** [reply]
Before asking you too many things. I was always curious about Translation Algorithms tab and Index reduction method.
Do you know bibliography about such things? I beleive that they are related about translating model files into equations and C. Am I right?
What about the jacobian methods? Do you know bibliography? 
Remember I am an engineer with basic numerical analysis courses.

Thanks for your knowledge

**Reply by casella** [reply]
I would highly recommend the book by François Cellier and Ernesto Kofman "Continuous System Simulation" Springer 2006.
